proof_of_stake/src/epoched.rs

```diff
@@ -368,6 +365,100 @@ where
         let key = self.get_last_update_storage_key();
         storage.write(&key, current_epoch)
     }
+
+    fn get_oldest_epoch_storage_key(&self) -> storage::Key {
+        self.storage_prefix
+            .push(&OLDEST_EPOCH_SUB_KEY.to_owned())
+            .unwrap()
+    }
+
+    fn get_oldest_epoch<S>(
+        &self,
+        storage: &S,
+    ) -> storage_api::Result<Option<Epoch>>
+    where
+        S: StorageRead,
+    {
+        let key = self.get_oldest_epoch_storage_key();
+        storage.read(&key)
+    }
+
+    fn set_oldest_epoch<S>(
+        &self,
+        storage: &mut S,
+        new_oldest_epoch: Epoch,
+    ) -> storage_api::Result<()>
+    where
+        S: StorageRead + StorageWrite,
+    {
+        let key = self.get_oldest_epoch_storage_key();
+        storage.write(&key, new_oldest_epoch)
+    }
+
+    fn sub_past_epochs(params: &PosParams, epoch: Epoch) -> Epoch {
+        epoch
+            .checked_sub(PastEpochs::value(params))
+            .unwrap_or_default()
+    }
+
+    /// Update data by removing old epochs
+    // TODO: should we consider more complex handling of empty epochs in the
+    // data below?
```

Comment:
What do you have in mind for this? I think for some `Epoched` types it's common not to have some entry at every past epoch (e.g. `ValidatorConsensusKeys`)

Reply:
Nothing in particular, just wanted to put this note here to revisit what I've implemented if necessary, since there was more complex handling in the non-nested impl. Do you think it's fine as is?

Reply:
I see, in the non-nested impl we set the latest value in case there isn't one at the oldest epoch to keep (`Self::sub_past_epochs`). For all the data where we're currently using `NestedEpoched` we don't need to do this as:
- `ValidatorSetPositions`, `ConsensusValidatorSets`, `BelowCapacityValidatorSets`, `ValidatorAddresses` have a value at every epoch that has not yet been cleared
- enqueued slashes in `EpochedSlashes` will be processed before they're cleared and processed slashes are not epoched

Let's just note this behavior in this rustdoc